### PR TITLE
Improve interface and fix issues - POC working end-to-end with Docker

### PR DESCRIPTION
**Pull request by n4ss**
The end-to-end POC is composed of:
- `libentitlement`: the branch used for this PR
- `Moby`: https://github.com/n4ss/moby/tree/poc-entitlements
- `docker CLI`: https://github.com/n4ss/cli/tree/entitlements-poc

Here is a changelist in this PR:
- Expose the network entitlements IDs (strings identifiers)
- Do not export entitlement enforcement callbacks
- Fix `network.none` entitlement to allow calls to `socket` for `Unix sockets`
- Fix  `network.user` entitlement to allow `bpf` and disallow setsockopt for `SO_DEBUG` sockets
- Add getter/setter to `EntitlementManager` for the security profile
- Add a function to enforce default entitlements on the security profile
- Add a function to allow a set of syscalls with their corresponding arguments in the seccomp profile
- Update vendored OCI spec version to Moby's

### CODE REVIEW

File: security-profile/security_profile.go [hunk 1]
```diff
@@ -84,8 +84,39 @@ func (p *Profile) AddNamespaces(nsTypes ...specs.LinuxNamespaceType) {
...
+	for syscallNameToAllow, syscallArgsToAllow := range syscallsWithArgsToAllow {
+
+		for _, syscallRule := range p.Oci.Linux.Seccomp.Syscalls {
+
+			if syscallRule.Action == specs.ActAllow {
+				for _, syscallName := range syscallRule.Names {
+					if syscallName == syscallNameToAllow &&
+						((len(syscallArgsToAllow) == 0 && len(syscallRule.Args) == 0) ||
+							reflect.DeepEqual(syscallRule.Args, syscallArgsToAllow)) {
+						return
```
Comment (riyazdf) [inline security-profile/security_profile.go]: should this be `continue`?  This function is hard to follow so I might be misunderstanding

**Comment by riyazdf** [comment]
should #20 be closed?

File: defaults/network-ents.go [hunk 2]
```diff
@@ -11,10 +12,17 @@ const (
...
 
 const (
-	NetworkNoneEntId  = "none"  // network.none
-	NetworkUserEntId  = "user"  // network.user
-	NetworkProxyEntId = "proxy" // network.proxy
-	NetworkAdminEntId = "admin" // network.admin
+	NetworkNoneEntFullId  = NetworkTLD + ".none"
+	NetworkUserEntFullId  = NetworkTLD + ".user"
+	NetworkProxyEntFullId = NetworkTLD + ".proxy"
+	NetworkAdminEntFullId = NetworkTLD + ".admin"
```
Comment (n4ss) [inline defaults/network-ents.go]: Will fix. does `networkDomain` works ?

File: defaults/network-ents.go [hunk 3]
```diff
@@ -42,6 +50,22 @@ func NetworkNoneEntitlement(profile *secProfile.Profile) (*secProfile.Profile, e
 	}
```
Comment (n4ss) [inline defaults/network-ents.go]: I'll modify the documentation sorry, docker needs these :)

File: libentitlement.go [hunk 4]
```diff
@@ -19,7 +20,7 @@ type EntitlementsManager struct {
 // default
 func NewEntitlementsManager(profile *secprofile.Profile) *EntitlementsManager {
 	if profile == nil {
-		logrus.Errorf("EntilementsManager initialization: invalid security profile - cannot be nil")
+		logrus.Errorf("Entilements Manager initialization: invalid security profile - cannot be nil")
```
Comment (n4ss) [inline libentitlement.go]: Will fix.

File: libentitlement.go [hunk 5]
```diff
@@ -31,8 +32,22 @@ func NewEntitlementsManager(profile *secprofile.Profile) *EntitlementsManager {
 }
 
 // GetProfile() returns the current state of the security profile
-func (m *EntitlementsManager) GetProfile() *secprofile.Profile {
-	return m.profile
+func (m *EntitlementsManager) GetProfile() (*secprofile.Profile, error) {
+	if m.profile == nil {
+		return nil, fmt.Errorf("Entitlements Manager has np security profile.")
```
Comment (n4ss) [inline libentitlement.go]: Will fix

File: libentitlement.go [hunk 6]
```diff
@@ -31,8 +32,22 @@ func NewEntitlementsManager(profile *secprofile.Profile) *EntitlementsManager {
...
+	}
+
+	return m.profile, nil
+}
+
+// SetProfile() sets the entitlement manager's security profile
```
Comment (n4ss) [inline libentitlement.go]: Yes, if the container management platform needs to:

1. add some entitlements -> init the entitlement manager with a profile
2. some configuration which may depend on profile infos -> get EntMgr profile
3. add more custom entitlements based on overall configuration -> set EntMgr profile
4. finish configuring

I'm not sure that all platforms want to do the entitlement configuration at the exact same moment in an atomic operation.

In Moby for example, there are already plenty of codepaths to set lots of defaults for masked paths, mount points etc.. and we won't be able necessarily to move all those codepaths right away to custom entitlements. This is why we need some way to get the profile from the entitlement manager, add stuff in it directly in Moby/Kub and then update the ent mgr profile.

File: libentitlement.go [hunk 7]
```diff
@@ -54,6 +69,16 @@ func isValidEntitlement(ent entitlement.Entitlement) (bool, error) {
 	return true, nil
 }
 
+// AddDefault() adds a default entitlement identified by the "
+func (m *EntitlementsManager) AddDefault(entName string) error {
+	defaultEnt, ok := defaults.DefaultEntitlements[entName]
+	if !ok {
+		return fmt.Errorf("Couldn't add invalid default entitlement name")
```
Comment (n4ss) [inline libentitlement.go]: nice catch!

File: security-profile/security_profile.go (hunk 1 above)
Comment (n4ss) [inline security-profile/security_profile.go]: Very nice catch thanks! 🎉

I'll add documentation also.

**Comment by n4ss** [comment]
@riyazdf I closed #20 but I'll add a test suite in a follow up PR, it's on my todo list.

File: libentitlement.go (hunk 6 above)
Comment (riyazdf) [inline libentitlement.go]: hm ok.  I was wondering if the output of 4. should be setting the entitlement manager to a "done" state and outputting a profile/some other spec.

We'll figure out what makes most sense after Moby/K8s integrations, so no need to over-specify now.

File: defaults/network-ents.go [hunk 8]
```diff
@@ -51,36 +75,35 @@ func NetworkNoneEntitlement(profile *secProfile.Profile) (*secProfile.Profile, e
...
-				ValueTwo: 0,
-				Op:       specs.OpEqualTo,
+				Index: 2,
+				Value: syscall.SO_DEBUG,
+				Op:    specs.OpNotEqual,
 			},
 		},
 	}
-	profile.BlockSyscallsWithArgs(syscallsWithArgsToBlock)
+	profile.AllowSyscallsWithArgs(syscallsWithArgsToBlock)
```
Comment (riyazdf) [inline defaults/network-ents.go]: can we rename the variable to `syscallsWithArgsToAllow`?

File: defaults/network-ents.go [hunk 9]
```diff
@@ -38,10 +46,26 @@ func NetworkNoneEntitlement(profile *secProfile.Profile) (*secProfile.Profile, e
...
+		"socket": {
+			{
+				Index: 0,
+				Op:    specs.OpEqualTo,
+				Value: syscall.AF_UNIX,
+			},
+			{
+				Index: 0,
+				Op:    specs.OpEqualTo,
+				Value: syscall.AF_LOCAL,
```
Comment (riyazdf) [inline defaults/network-ents.go]: nit: it would be nice to make note of these args in the function docstring, since this isn't a full block of `socket`

File: libentitlement.go [hunk 10]
```diff
@@ -19,7 +20,7 @@ type EntitlementsManager struct {
 // default
 func NewEntitlementsManager(profile *secprofile.Profile) *EntitlementsManager {
 	if profile == nil {
-		logrus.Errorf("EntilementsManager initialization: invalid security profile - cannot be nil")
+		logrus.Errorf("Entilement Manager initialization: invalid security profile - cannot be nil")
```
Comment (riyazdf) [inline libentitlement.go]: typo: Entitlement

File: defaults/network-ents.go (hunk 9 above)
Comment (n4ss) [inline defaults/network-ents.go]: Done

Comment (n4ss) [inline defaults/network-ents.go]: I also added doc for the remaining entitlements
